File: apps/ios/Sources/CrossyProtocol/Puzzle.swift
// The client puzzle (INV-6, DESIGN.md §4, §7; PROTOCOL.md §12). Twin of the client half
// of packages/protocol/src/puzzle.ts.
//
// INV-6 is structural here, deliberately more so than in the TS twin: this module
// defines NO `ServerPuzzle` and no `Solution` type at all. The solution-bearing shape
// exists only server-side; a client that cannot even spell the type cannot decode,
// hold, or re-encode a solution. INV6NoSolutionTests pins this with a reflection sweep
// and a decode-with-solution-present golden.
//
// PROTOCOL.md §12 pins the load-bearing fact (the solution split) and leaves the
// exhaustive puzzle schema (image clues, cross-references, per-cell numbering) to
// ingestion; this is the faithful minimal model the wire contract needs, matching
// `PuzzleBase` field for field.

/// One styled span of a clue's prose (clue-formatting wave, owner ruling 2026-07-12:
/// clue markup renders as structured runs, never stripped, never raw HTML). Twin of
/// the wire `ClueRun`: a piece of literal text `t` plus the styles `s` that wrap it.
///
/// Guarantees the server holds and the client relies on: the concatenation of a clue's
/// runs' `t` equals its plain `text`; runs are canonical (no empty runs, adjacent
/// equal-style runs pre-merged, style order "b","i","sub","sup"). Styles decode
/// tolerantly: an unknown style string is dropped by the mapper (forward compatibility,
/// so a newer server's style never breaks an older client), which is why `s` is
/// `[String]` here rather than an enum that would reject the unknown value and fail the
/// whole decode.
public struct ClueRun: Sendable, Equatable, Codable {
    /// The literal text of this span. The runs' `t` concatenate to the clue's `text`.
    public let t: String
    /// The styles wrapping this span ("i", "b", "sub", "sup"); absent when unstyled.
    /// Unknown strings are kept verbatim on decode so the mapper can ignore them,
    /// never rejected (a rejected value would fail the clue, contra the wave's rule).
    public let s: [String]

    public init(t: String, s: [String] = []) {
        self.t = t
        self.s = s
    }

    private enum CodingKeys: String, CodingKey { case t, s }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.t = try container.decode(String.self, forKey: .t)
        // `s` is optional-and-absent for an unstyled run (the wire omits it); a missing
        // or null field reads as no styles.
        self.s = try container.decodeIfPresent([String].self, forKey: .s) ?? []
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(t, forKey: .t)
        // Canonical form omits an empty style set, matching the wire (unstyled runs
        // carry no `s`), so a decode/encode round trip stays stable.
        if !s.isEmpty { try container.encode(s, forKey: .s) }
    }
}

/// A clue, structured at ingestion (DESIGN.md §7). No answer field, on either side of
/// the split. Twin of `Clue`.
///
/// `runs` is the additive clue-formatting field (owner ruling 2026-07-12): the styled
/// spelling of `text`. It is absent for unstyled clues and for every puzzle stored
/// before this feature, so plain `text` is the permanent fallback and the only field a
/// renderer ever needs. Decoding is deliberately tolerant: a missing or empty `runs`
/// means plain text, and a MALFORMED `runs` value (wrong shape, a bad element) also
/// falls back to nil rather than failing the clue, so one broken run can never sink a
/// whole puzzle decode. The custom `init(from:)` exists only for that swallow; the rest
/// of the shape stays the boring synthesized behavior.
public struct Clue: Sendable, Equatable, Codable {
    public let number: Int
    public let text: String
    public let cellIndices: [Int]
    /// The styled spelling of `text`, or nil for a plain clue. When present, the runs'
    /// `t` concatenate to `text` (the server's guarantee); when malformed on the wire,
    /// it decodes to nil and `text` carries the clue alone.
    public let runs: [ClueRun]?

    public init(number: Int, text: String, cellIndices: [Int], runs: [ClueRun]? = nil) {
        self.number = number
        self.text = text
        self.cellIndices = cellIndices
        self.runs = runs
    }

    private enum CodingKeys: String, CodingKey {
        case number, text, cellIndices, runs
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.number = try container.decode(Int.self, forKey: .number)
        self.text = try container.decode(String.self, forKey: .text)
        self.cellIndices = try container.decode([Int].self, forKey: .cellIndices)
        // Tolerant by design: a malformed `runs` must never sink the puzzle decode
        // (the wave's rule). `try?` swallows a wrong-shaped value to nil, leaving
        // `text` as the fallback; a truly absent field reads as nil too.
        self.runs = try? container.decodeIfPresent([ClueRun].self, forKey: .runs)
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(number, forKey: .number)
        try container.encode(text, forKey: .text)
        try container.encode(cellIndices, forKey: .cellIndices)
        // Absent stays off the wire (the omit-when-nil posture, matching shadedCircles).
        try container.encodeIfPresent(runs, forKey: .runs)
    }
}

/// Across and down clue lists. Twin of `Clues`.
public struct Clues: Sendable, Equatable, Codable {
    public let across: [Clue]
    public let down: [Clue]

    public init(across: [Clue], down: [Clue]) {
        self.across = across
        self.down = down
    }
}

/// The only puzzle type on any client-facing payload (REST §12). No solution field,
/// transitively (INV-6). Twin of `ClientPuzzle` (= `PuzzleBase`). `shadedCircles` is
/// genuinely optional-and-absent (the TS field is `?`; ingestion omits it when empty),
/// so synthesized Codable's decodeIfPresent/encodeIfPresent is exactly right and the
/// conformance stays synthesized.
public struct ClientPuzzle: Sendable, Equatable, Codable {
    public let rows: Int
    public let cols: Int
    /// Black-square cell indices: unplayable and immutable (PROTOCOL.md §4).
    public let blocks: [Int]
    /// Circled cell indices, a visual overlay (DESIGN.md §2).
    public let circles: [Int]
    /// Shaded-circle cells, a render variant of circles (DESIGN.md §7). Absent when none.
    public let shadedCircles: [Int]?
    public let clues: Clues

    public init(
        rows: Int,
        cols: Int,
        blocks: [Int],
        circles: [Int],
        shadedCircles: [Int]? = nil,
        clues: Clues
    ) {
        self.rows = rows
        self.cols = cols
        self.blocks = blocks
        self.circles = circles
        self.shadedCircles = shadedCircles
        self.clues = clues
    }
}

File: apps/android/protocol/src/main/kotlin/crossy/protocol/Puzzle.kt
// The client puzzle (INV-6, DESIGN.md §4, §7; PROTOCOL.md §12). Twin of the client half of
// packages/protocol/src/puzzle.ts and apps/ios Puzzle.swift.
//
// INV-6 is structural here, deliberately more so than in the TS twin: this module defines
// NO ServerPuzzle and no Solution type at all. The solution-bearing shape exists only
// server-side; a client that cannot even spell the type cannot decode, hold, or re-encode a
// solution. Inv6NoSolutionTests pins this with a descriptor sweep and a decode golden that
// shows a solution-bearing document cannot be represented.
//
// PROTOCOL.md §12 pins the load-bearing fact (the solution split) and leaves the exhaustive
// puzzle schema (image clues, cross-references, per-cell numbering) to ingestion; this is the
// faithful minimal model the wire contract needs, matching `PuzzleBase` field for field.

package crossy.protocol

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerializationException
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * One styled span of a clue's prose (PROTOCOL.md §12, clue-formatting wave: clue markup
 * renders as structured runs, never stripped, never raw HTML). Twin of `ClueRun`: literal
 * text `t` plus the styles `s` that wrap it.
 *
 * `s` decodes tolerantly: an unknown style string SURVIVES decode verbatim (a newer server's
 * style must never break an older client, so `s` is a `List<String>` rather than an enum that
 * would reject the value and fail the whole decode). It is absent-optional on the wire (an
 * unstyled run carries no `s`), so it defaults to empty and re-encodes off the wire when empty.
 */
@Serializable
public data class ClueRun(
    val t: String,
    val s: List<String> = emptyList(),
)

/**
 * A clue, structured at ingestion (DESIGN.md §7). No answer field, on either side of the
 * split. Twin of `Clue`.
 *
 * `runs` is the additive clue-formatting field: the styled spelling of `text`. It is absent
 * for unstyled clues and for every puzzle stored before the feature, so plain `text` is the
 * permanent fallback and the only field a renderer ever needs. Decoding is deliberately
 * tolerant: a missing or null `runs` means plain text, and a MALFORMED `runs` value (wrong
 * shape, a bad element, a non-string style) also falls back to null rather than failing the
 * clue, so one broken run can never sink a whole puzzle decode. ClueSerializer exists only for
 * that swallow; the field surface is otherwise the boring twin.
 */
@Serializable(with = ClueSerializer::class)
public data class Clue(
    val number: Int,
    val text: String,
    val cellIndices: List<Int>,
    val runs: List<ClueRun>? = null,
)

/**
 * The hand-written Clue codec. Everything but `runs` decodes as required; `runs` swallows a
 * malformed value to null (the wave's rule, twin of the Swift `try?`), so a broken run leaves
 * the clue on its `text` fallback and every sibling clue decodes cleanly. On encode an absent
 * `runs` stays off the wire (the omit-when-null posture).
 */
public object ClueSerializer : KSerializer<Clue> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor("crossy.protocol.Clue") {
            element<Int>("number")
            element<String>("text")
            element<List<Int>>("cellIndices")
            element<List<ClueRun>>("runs", isOptional = true)
        }

    override fun deserialize(decoder: Decoder): Clue {
        val input = decoder as? JsonDecoder
            ?: throw SerializationException("Clue decodes from JSON only")
        val obj = input.decodeJsonElement().jsonObject
        val number = obj["number"]?.jsonPrimitive?.int
            ?: throw SerializationException("clue.number is required")
        val text = obj["text"]?.jsonPrimitive?.content
            ?: throw SerializationException("clue.text is required")
        val cellIndices = (obj["cellIndices"]?.jsonArray
            ?: throw SerializationException("clue.cellIndices is required"))
            .map { it.jsonPrimitive.int }
        val runsElement = obj["runs"]
        val runs = if (runsElement == null || runsElement is JsonNull) {
            null
        } else {
            // Tolerant by design: a wrong-shaped runs, a run missing `t`, or a non-string
            // style must never sink the puzzle decode. runCatching swallows the failure to
            // null, leaving `text` as the fallback.
            runCatching {
                input.json.decodeFromJsonElement(ListSerializer(ClueRun.serializer()), runsElement)
            }.getOrNull()
        }
        return Clue(number, text, cellIndices, runs)
    }

    override fun serialize(encoder: Encoder, value: Clue) {
        val output = encoder as? JsonEncoder
            ?: throw SerializationException("Clue encodes to JSON only")
        val obj = buildJsonObject {
            put("number", value.number)
            put("text", value.text)
            put("cellIndices", buildJsonArray { value.cellIndices.forEach { add(JsonPrimitive(it)) } })
            // Absent stays off the wire (the omit-when-null posture, matching shadedCircles).
            value.runs?.let {
                put("runs", output.json.encodeToJsonElement(ListSerializer(ClueRun.serializer()), it))
            }
        }
        output.encodeJsonElement(obj)
    }
}

/** Across and down clue lists. Twin of `Clues`. */
@Serializable
public data class Clues(
    val across: List<Clue>,
    val down: List<Clue>,
)

/**
 * The only puzzle type on any client-facing payload (REST §12). No solution field,
 * transitively (INV-6). Twin of `ClientPuzzle` (= `PuzzleBase`). `shadedCircles` is genuinely
 * absent-optional (the TS field is `?`; ingestion omits it when empty), so a null default is
 * exactly right: an absent key decodes to null and a null re-encodes to an absent key.
 */
@Serializable
public data class ClientPuzzle(
    val rows: Int,
    val cols: Int,
    val blocks: List<Int>,
    val circles: List<Int>,
    val clues: Clues,
    val shadedCircles: List<Int>? = null,
)
